### PR TITLE
Provide interface to external generators that comply to TGenerator

### PR DESCRIPTION
**Pull request by preghenella**
This PR addresses the main points discussed in the JIRA ticket
https://alice.its.cern.ch/jira/browse/O2-386

The above-mentioned JIRA ticket tracks progress, requests and discussions about a new feature that aims at providing an interface to external event generators, such as the generators available in AliRoot (AliGenerators). This would enable one to run simulation with O2 software using as input event generator any of those available in AliRoot and in particular any event generator which is interfaced via the TGenerator ROOT class. This functionality allows one to basically have limitless flexibility in the definition and configuration of the event generator for O2 simulation.

The JIRA ticket also provides some basic information on how to use this interface for simple cases.

@sawenzel , can you please comment/review the code?

Cheers,
Roberto

### CODE REVIEW

File: Common/SimConfig/include/SimConfig/SimConfig.h [hunk 1]
```diff
@@ -26,6 +26,8 @@ struct SimConfigData {
   std::string mGenerator;                    // chosen VMC generator
   unsigned int mNEvents;                     // number of events to be simulated
   std::string mExtKinFileName;               // file name of external kinematics file (needed for ext kinematics generator)
+  std::string mExtGenFileName;               // file name of external kinematics file (needed for ext kinematics generator)
+  std::string mExtGenFuncName;               // file name of external kinematics file (needed for ext kinematics generator)
```
Comment (sawenzel) [inline Common/SimConfig/include/SimConfig/SimConfig.h]: please adjust the comments (no copy-paste)

File: Generators/include/Generators/Generator.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "FairGenerator.h"
+#include <vector>
+#include <array>
+
+namespace o2
+{
+namespace eventgen
+{
+
```
Comment (sawenzel) [inline Generators/include/Generators/Generator.h]: class needs clang formatting

File: Generators/include/Generators/Generator.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  /*****************************************************************/
+  /*****************************************************************/
+    
+  class Generator : public FairGenerator
```
Comment (sawenzel) [inline Generators/include/Generators/Generator.h]: could you add a few lines of docu what this class does (is supposed to provide). Is it a pure interface or does it have some base functionality? (Maybe then GeneratorBase or something would be better?)

File: Generators/include/Generators/PrimaryGenerator.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace o2
+{
+namespace eventgen
+{
+
+  /*****************************************************************/
+  /*****************************************************************/
+  
+  class PrimaryGenerator : public FairPrimaryGenerator
```
Comment (sawenzel) [inline Generators/include/Generators/PrimaryGenerator.h]: what is a PrimaryGenerator? What is the difference to "Generator"?

Comment (sawenzel) [reply Generators/include/Generators/PrimaryGenerator.h]: ok I see. This is mirroring the Fair(Primary)Generator semantics.

File: Generators/share/external/pythia6.C [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  };
+  std::string line, command;
+  while (std::getline(file, line)) {
+    /** remove comments **/
+    command = line.substr(0, line.find_first_of("#"));
+    if (command.length() == 0) continue;
+    py6->Pygive(command.c_str());
+  }
+  file.close();
+}
```
Comment (sawenzel) [inline Generators/share/external/pythia6.C]: could you add some docu on what are the constraints (template) for such macro files?
Another example interfacing with an AliRoot generator might also be useful.

File: Generators/src/Generator.cxx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    FairGenerator(name, title),
+    mBoost(0.)
+  {
+    /** constructor **/
+
+  }
+
+  /*****************************************************************/
+
+  Generator::~Generator()
```
Comment (sawenzel) [inline Generators/src/Generator.cxx]: consider using `= default` for such trivial destructors (constructors)

File: Generators/include/Generators/GeneratorTGenerator.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    /** default constructor **/
+    GeneratorTGenerator();
+    /** constructor with name and title **/
+    GeneratorTGenerator(const Char_t *name, const Char_t *title = "ALICEo2 TGenerator Generator");
+    /** destructor **/
+    virtual ~GeneratorTGenerator();
+
+    /** setters **/
+    void setGenerator(TGenerator *val) {mGenerator = val;};
```
Comment (sawenzel) [inline Generators/include/Generators/GeneratorTGenerator.h]: I suggest to call this `setTGenerator`

File: Generators/include/Generators/PrimaryGenerator.h (hunk 4 above)
Comment (preghenella) [reply Generators/include/Generators/PrimaryGenerator.h]: this header file was added to this PR by mistake, will be removed

File: Generators/include/Generators/Generator.h (hunk 3 above)
Comment (preghenella) [reply Generators/include/Generators/Generator.h]: it is a pure interface between the FairPrimaryGenerator and a generic event generator, supposed to define a protocol to be implemented by the concrete realisations of the interface. I'd rather keep the name Generator, but as you like

File: Generators/share/external/pythia6.C (hunk 5 above)
Comment (preghenella) [reply Generators/share/external/pythia6.C]: I have added `share/Generators/external/extgen.C` as a template function with information about the constraints for the macro files.

I have added `share/Generators/external/hijing.C` as an example to interface with `AliGenHijing` generator of AliRoot

**Comment by preghenella** [comment]
hello @sawenzel ,

there must be something I'm doing wrong in the `CMakeLisits.txt` to exclude the macros not to be checked. I'm still getting errors there

`The following tests FAILED:
	 57 - /Generators/share/external/hijing.C (Failed)
Errors while running CTest`

Thanks for input,
Roberto

**Comment by sawenzel** [comment]
@preghenella : I have updated the PR. One needs to add filenames to the ignore list. Should be ok now.

**Pull request merged**